src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -123,124 +129,94 @@
-            } catch (IOException | GeneralSecurityException e) {
-                Logger.e(TAG, "Failed to get private key from AndroidKeyStore", "",
-                        ADALError.ANDROIDKEYSTORE_FAILED, e);
-                throw e;
-            }
-        }
+        // load key for encryption if not loaded
+        loadSecretKeyForEncryption();
+        Logger.v(TAG, "Encrypt version:" + mBlobVersion);
+        final byte[] blobVersion = mBlobVersion.getBytes(AuthenticationConstants.ENCODING_UTF8);
+        final byte[] bytes = clearText.getBytes(AuthenticationConstants.ENCODING_UTF8);
 
-        return mKey;
-    }
+        // IV: Initialization vector that is needed to start CBC
+        byte[] iv = new byte[DATA_KEY_LENGTH];
+        mRandom.nextBytes(iv);
+        IvParameterSpec ivSpec = new IvParameterSpec(iv);
 
-    /**
-     * load key based on version for migration
-     * 
-     * @throws GeneralSecurityException
-     * @throws IOException
-     */
-    private SecretKey getKeyForVersion(String keyVersion) throws GeneralSecurityException, IOException {
+        // Set to encrypt mode
+        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
+        Mac mac = Mac.getInstance(MAC_ALGORITHM);
+        cipher.init(Cipher.ENCRYPT_MODE, mKey, ivSpec);
 
-        if (keyVersion.equals(VERSION_USER_DEFINED)) {
-            return getSecretKey(AuthenticationSettings.INSTANCE.getSecretKeyData());
-        }
+        byte[] encrypted = cipher.doFinal(bytes);
 
-        if (keyVersion.equals(VERSION_ANDROID_KEY_STORE)) {
-            if (Build.VERSION.SDK_INT >= 18) {
-                try {
-                    // androidKeyStore can store app specific self signed cert.
-                    // Asymmetric cryptography is used to protect the session
-                    // key
-                    // used for Encryption and HMac
-                    return getSecretKeyFromAndroidKeyStore();
-                } catch (IOException | GeneralSecurityException e) {
-                    Logger.e(TAG, "Failed to get private key from AndroidKeyStore", "",
-                            ADALError.ANDROIDKEYSTORE_FAILED, e);
-                    throw e;
-                }
-            } else {
-                throw new IllegalArgumentException(
-                        String.format(
-                                "keyVersion '%s' is not supported in this SDK. AndroidKeyStore is supported API18 and above.",
-                                keyVersion));
-            }
-        }
+        // Mac output to sign encryptedData+IV. Keyversion is not included
+        // in the digest. It defines what to use for Mac Key.
+        mac.init(mMacKey);
+        mac.update(blobVersion);
+        mac.update(encrypted);
+        mac.update(iv);
+        byte[] macDigest = mac.doFinal();
 
-        throw new IllegalArgumentException("keyVersion = " + keyVersion);
-    }
+        // Init array to store blobVersion, encrypted data, iv, macdigest
```

Comment:
Plan to put this into a object. i.e. EncrpyBlobContainer. But will be in a separate PR.